controller/controller.go

```diff
@@ -112,6 +113,38 @@ type Controller struct {
 	Interval time.Duration
 	// The DomainFilter defines which DNS records to keep or exclude
 	DomainFilter endpoint.DomainFilter
+	// Used by RunOnceThrottled to ensure throttling, once per Interval
+	running      bool
+	minStartTime time.Time
+}
+
+var nowValue = struct{}{}
+
+// Make sure execution happens at most once per interval
+func (c *Controller) RunOnceThrottled(ctx context.Context) error {
+	now := time.Now()
+	testing := false
+	if v, ok := ctx.Value(nowValue).(time.Time); ok {
+		now = v
+		testing = true
```

Comment:
What about having an early return here by testing the condition on line 131 instead of having multiple if statements like in line 144?

Reply:
Both ifs are on purpose the one on line 134 represents case when function is throttled, the one on line 144 represents case when function executes normally. We need both to properly test this function in tests. Please note that in one case error is returned and in other nil (error returned in tests means that function is trottled, nil returned means no throttling).